### PR TITLE
Gal 276/data merge gridall imd

### PR DESCRIPTION
**Pull request by zainm97**
## Description
Ticket: https://nhsd-jira.digital.nhs.uk/browse/GAL-276

This PR combines the data from the IMD file with gridall records to generate a complete postcode record with location and, depravity ranking and score. 

## Type of changes

- [ ] Refactoring (non-breaking change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would change existing functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)

## Checklist

- [ ] I am familiar with the [contributing guidelines](../docs/CONTRIBUTING.md)
- [ ] I have followed the code style of the project
- [ ] I have added tests to cover my changes
- [ ] I have updated the documentation accordingly
- [ ] This PR is a result of pair or mob programming

---

## Sensitive Information Declaration

To ensure the utmost confidentiality and protect your and others privacy, we kindly ask you to NOT including [PII (Personal Identifiable Information) / PID (Personal Identifiable Data)](https://digital.nhs.uk/data-and-information/keeping-data-safe-and-benefitting-the-public) or any other sensitive data in this PR (Pull Request) and the codebase changes. We will remove any PR that do contain any sensitive information. We really appreciate your cooperation in this matter.

- [ ] I confirm that neither PII/PID nor sensitive data are included in this PR and the codebase changes.

### CODE REVIEW

**Review by joad432 — APPROVED** [review]
Nice work Zain

**Comment by joad432** [comment]
Everything looks good

File: terraform/environment/dev/resources/lambda/filterData/lambdaHandler/dataFilterLambda.js [hunk 1]
```diff
@@ -115,12 +115,46 @@ export const generateCsvString = (header, dataArray) => {
...
+    }
+  })
+  console.log('Function to combine records took: ', ((Date.now() - startTimeMatching)/1000)/60);
+  console.log('Amount of combined records: ' , count);
+  return lsoaArray;
+}
+
 // Lambda Entry Point
 export const handler = async () => {
   const bucketName = "galleri-ons-data";
```
Comment (LaithM96) [inline terraform/environment/dev/resources/lambda/filterData/lambdaHandler/dataFilterLambda.js:150]: "Consider using environment variables (e.g., process.env.BUCKET_NAME) for better security and flexibility."

File: terraform/environment/dev/resources/lambda/tests/unitTests/dataFilterLambda.test.js [hunk 2]
```diff
@@ -11,7 +11,8 @@ import {
   processGridallRow,
   processImdRow,
   generateCsvString,
-  parseCsvToArray
+  parseCsvToArray,
+  mergeImdGridallData
 } from '../../filterData/lambdaHandler/dataFilterLambda.js';
 
 describe("readCsvFromS3", () => {
```
Comment (LaithM96) [inline terraform/environment/dev/resources/lambda/tests/unitTests/dataFilterLambda.test.js:18]: "Ensure all necessary cleanup operations are performed in the afterEach or afterAll hooks."

File: terraform/environment/dev/resources/lambda/filterData/lambdaHandler/dataFilterLambda.js [hunk 3]
```diff
@@ -115,12 +115,46 @@ export const generateCsvString = (header, dataArray) => {
...
+      return gridallRecord;
+    }
+  })
+  console.log('Function to combine records took: ', ((Date.now() - startTimeMatching)/1000)/60);
+  console.log('Amount of combined records: ' , count);
+  return lsoaArray;
+}
+
 // Lambda Entry Point
 export const handler = async () => {
```
Comment (LaithM96) [inline terraform/environment/dev/resources/lambda/filterData/lambdaHandler/dataFilterLambda.js:149]: This function is quite lengthy. Consider breaking it into smaller functions or modules for better maintainability."

File: terraform/environment/dev/resources/lambda/filterData/lambdaHandler/dataFilterLambda.js [hunk 4]
```diff
@@ -115,12 +115,46 @@ export const generateCsvString = (header, dataArray) => {
   ].join("\n");
 };
 
+export const mergeImdGridallData = (gridallData, imdData, startTime) => {
+  console.log('Attempting to format imd dictionary records')
+  let imdDict = {};
```
Comment (LaithM96) [inline terraform/environment/dev/resources/lambda/filterData/lambdaHandler/dataFilterLambda.js:120]: Ensure this dictionary doesn't grow too large, which might cause memory issues. Consider optimisations if the dataset is large.

**Review by LaithM96 — CHANGES_REQUESTED** [review]
- [x] Ensure no sensitive data or secrets are hardcoded in the code. Use environment variables or AWS Secrets Manager.
- [ ] Consider adding more descriptive comments before complex functions or logic for better readability.
- [ ] Consider version pinning for your AWS provider to ensure consistent behaviour.
- [ ] Consider adding more edge cases and negative test scenarios to ensure comprehensive coverage.

but overall looks good, well done 🥇

File: terraform/environment/dev/resources/lambda/filterData/lambdaHandler/dataFilterLambda.js (hunk 3 above)
Comment (zainm97) [reply terraform/environment/dev/resources/lambda/filterData/lambdaHandler/dataFilterLambda.js:149]: The logic should be contained in one function imo. Its also 16 lines so not too large. As a rule of thumb if code is longer than 50 lines it's usually an indication it could be made into a function

File: terraform/environment/dev/resources/lambda/filterData/lambdaHandler/dataFilterLambda.js (hunk 4 above)
Comment (zainm97) [reply terraform/environment/dev/resources/lambda/filterData/lambdaHandler/dataFilterLambda.js:120]: Dictionary is the most efficient way i could think of dealing with the data. They have O(1) time complexity for look up and retrieval. The imd file is constant and we don't anticipate it changing. You can also look at the logs as i have timed the operations. This function completes in the order of seconds

File: terraform/environment/dev/resources/lambda/tests/unitTests/dataFilterLambda.test.js (hunk 2 above)
Comment (zainm97) [reply terraform/environment/dev/resources/lambda/tests/unitTests/dataFilterLambda.test.js:18]: yup, mocks cleared up as evidenced in line 19. All other variables go out of scope when the particular test() function exits

File: terraform/environment/dev/resources/lambda/filterData/lambdaHandler/dataFilterLambda.js (hunk 1 above)
Comment (zainm97) [reply terraform/environment/dev/resources/lambda/filterData/lambdaHandler/dataFilterLambda.js]: this is a good point. Actioning it now

**Review by LaithM96 — APPROVED** [review]
from the comment and the changes done, happy to approve, good work 💯

**Review by msandhuNHS — APPROVED** [review]
